File: common/src/main/java/com/dreamdisplays/ytdlp/FormatDiskCache.java
package com.dreamdisplays.ytdlp;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.google.gson.reflect.TypeToken;
import me.inotsleep.utils.logging.LoggingManager;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

import java.io.File;
import java.io.IOException;
import java.lang.reflect.Type;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.HexFormat;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Persistent on-disk cache for resolved YouTube format URLs. YouTube signed URLs are
 * typically valid for ~6 hours, so caching across sessions removes the cold-start
 * yt-dlp invocation when a player rejoins a server with the same displays.
 */
@NullMarked
public final class FormatDiskCache {

    private static final Path CACHE_DIR = Path.of("config", "dreamdisplays", "yt-cache");
    private static final Gson GSON = new GsonBuilder().create();
    private static final Type STREAM_LIST_TYPE = new TypeToken<List<YtStream>>() {
    }.getType();
    private static final long DEFAULT_TTL_MS = 5L * 60L * 60L * 1_000L; // 5h

    private static final ExecutorService WRITER = Executors.newSingleThreadExecutor(r -> {
        Thread t = new Thread(r, "DD-FormatCache-writer");
        t.setDaemon(true);
        return t;
    });

    private FormatDiskCache() {
    }

    public static @Nullable List<YtStream> load(String videoUrl, long maxAgeMs) {
        File f = fileFor(videoUrl);
        if (!f.isFile()) return null;
        try {
            String json = Files.readString(f.toPath(), StandardCharsets.UTF_8);
            JsonObject obj = JsonParser.parseString(json).getAsJsonObject();
            long ts = obj.has("ts") ? obj.get("ts").getAsLong() : 0L;
            if (System.currentTimeMillis() - ts > maxAgeMs) {
                f.delete();
                return null;
            }
            List<YtStream> streams = GSON.fromJson(obj.get("streams"), STREAM_LIST_TYPE);
            return streams == null || streams.isEmpty() ? null : streams;
        } catch (Exception e) {
            // Treat corrupted entries as cache miss
            try {
                f.delete();
            } catch (Exception ignored) {
            }
            return null;
        }
    }

    public static @Nullable List<YtStream> load(String videoUrl) {
        return load(videoUrl, DEFAULT_TTL_MS);
    }

    public static void saveAsync(String videoUrl, List<YtStream> streams) {
        if (streams.isEmpty()) return;
        WRITER.submit(() -> writeNow(videoUrl, streams));
    }

    private static void writeNow(String videoUrl, List<YtStream> streams) {
        try {
            Files.createDirectories(CACHE_DIR);
            File target = fileFor(videoUrl);
            File tmp = new File(target.getParentFile(), target.getName() + ".tmp");
            JsonObject root = new JsonObject();
            root.addProperty("ts", System.currentTimeMillis());
            root.addProperty("url", videoUrl);
            root.add("streams", GSON.toJsonTree(streams, STREAM_LIST_TYPE));
            Files.writeString(tmp.toPath(), GSON.toJson(root), StandardCharsets.UTF_8);
            Files.move(tmp.toPath(), target.toPath(),
                    java.nio.file.StandardCopyOption.REPLACE_EXISTING,
                    java.nio.file.StandardCopyOption.ATOMIC_MOVE);
        } catch (IOException e) {
            LoggingManager.warn("FormatDiskCache write failed: " + e.getMessage());
        }
    }

    public static void sweepExpired() {
        sweepExpired(DEFAULT_TTL_MS);
    }

    public static void sweepExpired(long maxAgeMs) {
        try {
            if (!Files.isDirectory(CACHE_DIR)) return;
            long now = System.currentTimeMillis();
            try (var stream = Files.list(CACHE_DIR)) {
                stream.filter(p -> p.toString().endsWith(".json")).forEach(p -> {
                    try {
                        String json = Files.readString(p, StandardCharsets.UTF_8);
                        JsonObject obj = JsonParser.parseString(json).getAsJsonObject();
                        long ts = obj.has("ts") ? obj.get("ts").getAsLong() : 0L;
                        if (now - ts > maxAgeMs) Files.deleteIfExists(p);
                    } catch (Exception ignored) {
                        try { Files.deleteIfExists(p); } catch (Exception ignored2) {}
                    }
                });
            }
        } catch (IOException ignored) {
        }
    }

    private static File fileFor(String videoUrl) {
        return new File(CACHE_DIR.toFile(), hash(videoUrl) + ".json");
    }

    private static String hash(String s) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-1");
            return HexFormat.of().formatHex(md.digest(s.getBytes(StandardCharsets.UTF_8)));
        } catch (NoSuchAlgorithmException e) {
            return Integer.toHexString(s.hashCode());
        }
    }
}

File: common/src/main/kotlin/com/dreamdisplays/ytdlp/FormatDiskCache.kt
package com.dreamdisplays.ytdlp

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.google.gson.reflect.TypeToken
import me.inotsleep.utils.logging.LoggingManager
import java.io.File
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.util.*
import java.util.concurrent.Executors

/**
 * Persistent on-disk cache for resolved YouTube format URLs.
 */
object FormatDiskCache {
    private val CACHE_DIR: Path = Path.of("config", "dreamdisplays", "yt-cache")
    private val GSON: Gson = GsonBuilder().create()
    private val STREAM_LIST_TYPE = object : TypeToken<List<YtStream>>() {}.type
    private const val DEFAULT_TTL_MS = 5L * 60L * 60L * 1_000L
    private const val SCHEMA_VERSION = 2

    private val WRITER = Executors.newSingleThreadExecutor { r ->
        Thread(r, "DD-FormatCache-writer").apply { isDaemon = true }
    }

    /** Reads the cached streams for [videoUrl] from disk; returns null if absent, expired, or schema-mismatched. */
    fun load(videoUrl: String, maxAgeMs: Long): List<YtStream>? {
        val f = fileFor(videoUrl)
        if (!f.isFile) return null
        return try {
            val json = Files.readString(f.toPath(), StandardCharsets.UTF_8)
            val obj = JsonParser.parseString(json).asJsonObject
            val version = if (obj.has("v")) obj.get("v").asInt else 0
            if (version != SCHEMA_VERSION) {
                f.delete()
                return null
            }
            val ts = if (obj.has("ts")) obj.get("ts").asLong else 0L
            if (System.currentTimeMillis() - ts > maxAgeMs) {
                f.delete()
                return null
            }
            val streams: List<YtStream>? = GSON.fromJson(obj.get("streams"), STREAM_LIST_TYPE)
            if (streams.isNullOrEmpty()) null else streams
        } catch (_: Exception) {
            runCatching { f.delete() }
            null
        }
    }

    /** Serialises [streams] to disk for [videoUrl] asynchronously via the single-threaded writer executor. */
    fun saveAsync(videoUrl: String, streams: List<YtStream>) {
        if (streams.isEmpty()) return
        WRITER.submit { writeNow(videoUrl, streams) }
    }

    /** Atomically writes the stream JSON to disk using a temp file and rename. */
    private fun writeNow(videoUrl: String, streams: List<YtStream>) {
        try {
            Files.createDirectories(CACHE_DIR)
            val target = fileFor(videoUrl)
            val tmp = File(target.parentFile, target.name + ".tmp")
            val root = JsonObject().apply {
                addProperty("v", SCHEMA_VERSION)
                addProperty("ts", System.currentTimeMillis())
                addProperty("url", videoUrl)
                add("streams", GSON.toJsonTree(streams, STREAM_LIST_TYPE))
            }
            Files.writeString(tmp.toPath(), GSON.toJson(root), StandardCharsets.UTF_8)
            Files.move(
                tmp.toPath(), target.toPath(),
                StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE
            )
        } catch (e: IOException) {
            LoggingManager.warn("[FormatDiskCache] Write failed: ${e.message}")
        }
    }

    /** Deletes the cache entry for [videoUrl] asynchronously. */
    fun deleteEntry(videoUrl: String) {
        WRITER.submit {
            runCatching { Files.deleteIfExists(fileFor(videoUrl).toPath()) }
        }
    }

    /** Scans the cache directory and deletes all `.json` entries older than [maxAgeMs] milliseconds. */
    fun sweepExpired(maxAgeMs: Long = DEFAULT_TTL_MS) {
        try {
            if (!Files.isDirectory(CACHE_DIR)) return
            val now = System.currentTimeMillis()
            Files.list(CACHE_DIR).use { stream ->
                stream.filter { it.toString().endsWith(".json") }.forEach { p ->
                    try {
                        val json = Files.readString(p, StandardCharsets.UTF_8)
                        val obj = JsonParser.parseString(json).asJsonObject
                        val ts = if (obj.has("ts")) obj.get("ts").asLong else 0L
                        if (now - ts > maxAgeMs) Files.deleteIfExists(p)
                    } catch (_: Exception) {
                        runCatching { Files.deleteIfExists(p) }
                    }
                }
            }
        } catch (_: IOException) {
        }
    }

    /** Returns the cache file path for [videoUrl] by hashing the URL to a stable filename. */
    private fun fileFor(videoUrl: String): File = File(CACHE_DIR.toFile(), hash(videoUrl) + ".json")

    /** Returns a SHA-1 hex digest of [s], falling back to `hashCode` if SHA-1 is unavailable. */
    private fun hash(s: String): String = try {
        val md = MessageDigest.getInstance("SHA-1")
        HexFormat.of().formatHex(md.digest(s.toByteArray(StandardCharsets.UTF_8)))
    } catch (_: NoSuchAlgorithmException) {
        Integer.toHexString(s.hashCode())
    }
}
